How tos - DOC-314 (#472)
matter-labs/zksync-web-era-docs: docs/dev/developer-guides/bridging/l1-l2-interop.md

# L1 / L2 interoperability

## Common use cases

Many use cases require multi-layer interoperability, such as:

- The network's censorship resistance.
- Custom bridges.
- Multi-layer governing smart contracts.
- Multi-layer transfers.

## L1 to L2 communication

L1 to L2 communication is governed by the [`IZkSync.sol`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IZkSync.sol#L4) inherited interfaces.

:::tip
- If you prefer to learn-by-doing, the [cross chain governance tutorial](../../tutorials/cross-chain-tutorial.md) is a practical example of layer interoperability.
:::

### Gas estimation

> The system processes gas estimation for transactions implicitly. However, it is also possible to implement the gas estimation processes explicitly.

Reviewer: Maybe a matter of taste, but I call zkSync "the system", so the system does not calculate the limit implicitly. The SDK does that

```suggestion
The SDK processes gas estimation for transactions implicitly. However, it is also possible to implement the gas estimation processes explicitly.
```

:::tip L1 to L2 gas estimation for transactions
- Basic costs are measured in the amount of gas, and so the final cost depends on the gas price that the transaction assigns.
- The transaction process requires the current L1 gas price, transaction base cost, and transaction gas limit which defines the maximum amount of gas a transaction can consume.
:::

- Find out [how to estimate gas](../../how-to/estimate-gas.md) for different scenarios.
- Find out [how to send a transaction from L1 to L2](../../how-to/send-transaction-l1-l2.md) with zkSync Era.

## L2 to L1 

L2 to L1 communication is based on transferring the data as a message, and not on L1 transaction execution. 

- Find out [how to send a message from L2 to L1](../../how-to/send-message-l2-l1.md) with zkSync Era.

## Priority queue

> 1. All transactions types are now supported by the priority queue. 

Reviewer 2: ```suggestion
1. All transactions types are supported by the priority queue.
```

I presume the "now" was there in comparison to v1.

:::tip L1 to L2 gas estimation for transactions
- Basic costs are measured in the amount of gas, and so the final cost depends on the gas price that the transaction assigns.
- The transaction process requires the current L1 gas price, transaction base cost, and transaction gas limit which defines the maximum amount of gas a transaction can consume.
:::

- Find out [how to estimate gas](../../how-to/estimate-gas.md) for different scenarios.
- Find out [how to send a transaction from L1 to L2](../../how-to/send-transaction-l1-l2.md) with zkSync Era.

## L2 to L1 

L2 to L1 communication is based on transferring the data as a message, and not on L1 transaction execution. 

- Find out [how to send a message from L2 to L1](../../how-to/send-message-l2-l1.md) with zkSync Era.

## Priority queue

1. All transactions types are now supported by the priority queue. 

    In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

    :::info

>     The withdraw functionality in zkSync Era replaces `FullExit` in Lite.

Reviewer 2: (considering my last comment) here's another mention of (zkSync) Lite. I also believe that what is written is incorrect -- both versions have a "withdraw" transaction type (which does exactly the same thing)

Author: that also got removed in the last commit

# L1 / L2 interoperability

## Common use cases

Many use cases require multi-layer interoperability, such as:

- The network's censorship resistance.
- Custom bridges.
- Multi-layer governing smart contracts.
- Multi-layer transfers.

## L1 to L2 communication

L1 to L2 communication is governed by the [`IZkSync.sol`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IZkSync.sol#L4) inherited interfaces.

:::tip
- If you prefer to learn-by-doing, the [cross chain governance tutorial](../../tutorials/cross-chain-tutorial.md) is a practical example of layer interoperability.
:::

### Gas estimation

The system processes gas estimation for transactions implicitly. However, it is also possible to implement the gas estimation processes explicitly.

:::tip L1 to L2 gas estimation for transactions

> - Gas is measured in amount and not ERC20 token value so actual costs can vary.

Reviewer: > Gas is measured in amount and not ERC20 token value so actual costs can vary

I am not sure what this means

Author: amount == number of gas

Reviewer 2: To remind you that I also had an issue with this phrasing. Maybe better to modify

Author: try now

> - The transaction process requires the current L1 gas price, transaction base cost, and transaction gas limit which defines the minimum amount of gas a transaction requires.

Reviewer: I don't really understand what this sentence means, sorry

Author: a tx requires L1 gas price, base cost, and gas limit ...

Author: do you have any problem with this one @barakshani ?

Reviewer 2: Only that I think this is redundant : ) (see [here](https://github.com/matter-labs/zksync-web-era-docs/pull/472#discussion_r1196260364))

But technically speaking transactions only need the base cost and gas limit. If it just happens that the base cost is lower than the current L1 gas price the tx won't be processed until it is flipped.

Author: Does it work in a different way than Ethereum? If so, then it is incorrect. Is the redundancy the assumption that the reader knows Ethereum inside out? I don't think that's a safe assumption.

Author: @StanislavBreadless anything more on this one?

:::

- Find out [how to estimate gas](../../how-to/estimate-gas.md) for different scenarios.
- Find out [how to send a transaction from L1 to L2](../../how-to/send-transaction-l1-l2.md) with zkSync Era.

## L2 to L1 

L2 to L1 communication is based on transferring the data as a message, and not on L1 transaction execution. 

- Find out [how to send a message from L2 to L1](../../how-to/send-message-l2-l1.md) with zkSync Era.

## Priority queue

> zkSync Era supports general smart contract computation which includes complex operations. For this reason, some principles had to evolve in order to preserve the stability of the network.

Reviewer: I think that when we use `evolve` we need to show compared to what the evolution is done, I guess a lot of people that make come to version 2 have on prior experience with version 1

Author: removed

1. All transactions types are now supported by the priority queue. 

    In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

    :::info
    The withdraw functionality in zkSync Era replaces `FullExit` in Lite.
    :::

2. The priority queue must be fully permissionless to prevent malicious activity. For example, malicious users might send multiple transactions which push up the block gas limit to unworkable levels. To mitigate against this, submitting transactions to the priority queue is no longer free and users must pay a fee to the operator. 

>     The fee for a transaction is equal to `txBaseCost * gasPrice`. The `gasPrice` is the gas price of the user's transaction, while `txBaseCost` is the base cost for the transaction and depends on the transaction parameters (e.g. `gasLimit` for an `Execute` transaction).

Reviewer: The fee is equal to `txBaseCost`

Author: this whole paragraph is now dubious and needs rewriting (by the way, this has been live for at least [7 months](https://github.com/matter-labs/zksync-web-era-docs/blob/ad8d4de9d9415cbe5db4befb0bcd7487875ed373/docs/dev/developer-guides/bridging/l1-l2-interop.md))

Author: Also, here @barakshani ?

Reviewer 2: I'm not sure what exactly txBaseCost stands for (it's does not exist on L1), Stas should know better.

Reviewer: @kailash-manasarovar we call the `baseCost` the minimal amount of ETH that needs to be paid for the L1->L2 transaction to get into zkSync

Author: Can someone rewrite the paragraph so it is correct please. I am not able to do this.

Reviewer 2: Considering the context, this is what I would write:
"To obtain the cost for sending an L2 to L1 message, please refer [here](https://era.zksync.io/docs/dev/developer-guides/bridging/l1-l2.html#getting-the-base-cost)."

Author: @barakshani to be very clear, you are saying, remove this:

"The fee for a transaction is equal to `txBaseCost * gasPrice`. The `gasPrice` is the gas price of the user's transaction, while `txBaseCost` is the base cost for the transaction and depends on the transaction parameters (e.g. `gasLimit` for an `Execute` transaction)."

In favour of this:

"To obtain the cost for sending an L2 to L1 message, please refer [here](https://era.zksync.io/docs/dev/developer-guides/bridging/l1-l2.html#getting-the-base-cost)."

?

Reviewer 2: I do

Author: OK, I have just made this change and it now has no relation to the priority queue section at all. It is like a disembodied sentence.

"
## Priority queue

1. All transactions types are now supported by the priority queue. 

2. The priority queue must be fully permissionless to prevent malicious activity. For example, malicious users might send multiple transactions which push up the block gas limit to unworkable levels. To mitigate against this, submitting transactions to the priority queue is no longer free and users must pay a fee to the operator. 

To obtain the cost for sending an L2 to L1 message, please refer to [step 4 of how to send an L1 to L2 transaction](../../how-to/send-transaction-l1-l2.md#step-by-step).
"

On this page, we already have links to these instructions so I'm just going to remove this sentence and we'll have point 1, point 2, and that's it.

Author: please check latest commit -> also some of the other please from yesterday

Reviewer 2: ok I see -- in that case we can put the sentence under point 2. It relates to the cost that are mentioned there.

Author: ok added back

:::tip
- If you prefer to learn-by-doing, the [cross chain governance tutorial](../../tutorials/cross-chain-tutorial.md) is a practical example of layer interoperability.
:::

### Gas estimation

The system processes gas estimation for transactions implicitly. However, it is also possible to implement the gas estimation processes explicitly.

:::tip L1 to L2 gas estimation for transactions
- Gas is measured in amount and not ERC20 token value so actual costs can vary.
- The transaction process requires the current L1 gas price, transaction base cost, and transaction gas limit which defines the minimum amount of gas a transaction requires.
:::

- Find out [how to estimate gas](../../how-to/estimate-gas.md) for different scenarios.
- Find out [how to send a transaction from L1 to L2](../../how-to/send-transaction-l1-l2.md) with zkSync Era.

## L2 to L1 

L2 to L1 communication is based on transferring the data as a message, and not on L1 transaction execution. 

- Find out [how to send a message from L2 to L1](../../how-to/send-message-l2-l1.md) with zkSync Era.

## Priority queue

> zkSync Lite is sufficient for a system that only supports deposit and full exit operations.

Reviewer 2: You want to say "The priority queue in zkSync Lite..." 

But this sentence is not correct. zkSync lite supports token transfers and atomic swaps as well. [but the priority queue does not support these, as it was not design to address these]

Either find a better way of relating to Lite, or just remove this bit per favor

Author: removed

# L1 / L2 interoperability

## Common use cases

Many use cases require multi-layer interoperability, such as:

- The network's censorship resistance.
- Custom bridges.
- Multi-layer governing smart contracts.
- Multi-layer transfers.

## L1 to L2 communication

L1 to L2 communication is governed by the [`IZkSync.sol`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IZkSync.sol#L4) inherited interfaces.

:::tip
- If you prefer to learn-by-doing, the [cross chain governance tutorial](../../tutorials/cross-chain-tutorial.md) is a practical example of layer interoperability.
:::

### Gas estimation

The system processes gas estimation for transactions implicitly. However, it is also possible to implement the gas estimation processes explicitly.

:::tip L1 to L2 gas estimation for transactions

> - Gas is measured in amount and not ERC20 token value so actual costs can vary.

Reviewer 2: I don't get this comment

Author: Amount of gas versus value of gas, i.e. 10 beans ... instead of £1.50 per bean = £15.

Reviewer 2: 1. why do we even mention erc20 token value?
2. as I wrote in the subsequent comment, this is just how Ethereum works. Not sure why we need to explicitly state it.

Author: Because I was not clear whether it was value or amount - as the reader. And I expect a newbie would be the same.

Reviewer 2: ok, I just find "ERC20 token value" kinda confusing, but if you want to stress it, I don't have a better suggestion (we can use "monetary value" but I'm not sure it's better)

Author: let's put it on the back-burner and if we come up with something better... we'll add it wdyt?

> - The transaction process requires the current L1 gas price, transaction base cost, and transaction gas limit which defines the minimum amount of gas a transaction requires.

Reviewer 2: This, and the previous sentence, seem to simply state how things work on Ethereum. Is this the intent?

> gas limit which defines the minimum amount of gas a transaction requires.

gas limit limits the amount of gas the transaction will consume. So it's about the maximum gas, not the minimum

Author: I also thought maximum @barakshani but it is minimum @StanislavBreadless please confirm.

Reviewer 2: @kailash-manasarovar please correct (here and in other places): gas limit isn't the minimum

Author: Wait, I thought we established that it was indeed the lower bound and than consequently additional values are added to it to ensure fee coverage?

Author: This is really tricky, when one dev disagrees with another on something so fundamental like this.

Reviewer 2: @StanislavBreadless do you disagree with anything in my previous comment?

And the context is whether we should state that the gas estimation or gas limit is a limit from below

Reviewer: @barakshani 

It attempts to return the minimal value which is required for the transaction to execute successfully. Sometimes we do not know for sure and return an estimate or roughly minimal value. If it did not at least try to return the minimal value for the transaction to succeed, we could always supply maximal possible gas limit : ) 

I understand that both you and me know what `gasLimit` is, I don't mind anything if the wording feels better for you

Author: your thoughts @barakshani .. just to say, I originally had it saying maximum, and it was technically edited by SME input to minimum, and I'm still struggling with minimal == maximum ...

Reviewer 2: Again, there are 2 different values we discuss here:
- The answer of estimageGas
- gasLimit within the tx

The dispute is about the first, where I claim that no "minimum" (nor maximum) should be used in this context (also, if it was the minimum we wouldn't have "out of gas" errors, that shows that the tx needed more gas).

We can refrain from speaking about minimum/maximum, but we cannot say that gasLimit is the minimum that the tx "requires" -- it is not, and it is the maximum the tx can consume. There's no disagreement between the two of us about it.

Reviewer 2: lgtm. thanks!

:::

- Find out [how to estimate gas](../../how-to/estimate-gas.md) for different scenarios.
- Find out [how to send a transaction from L1 to L2](../../how-to/send-transaction-l1-l2.md) with zkSync Era.

### Priority queue

> The priority queue provides a censorship-resistant way to interact with zkSync Era in case the operator becomes malicious or unavailable.

Reviewer 2: "permissionless" is a blockchain term.

I have another comment where censorship-resistant might fit better, but I do think that "permissionless way" is more correct, and as I explain, if the operator is unavailable then being censorship-resistant won't help in this case (but being permissionless does).

> The priority queue provides a censorship-resistant way to interact with zkSync Era in case the operator becomes malicious or unavailable.

Author: ok, amended

The priority queue functions in a similar way it did in the previous version of zkSync. For the full picture, we explain the priority queue on zkSync Lite before detailing the new design in zkSync Era.

> #### How it worked in zkSync Lite

Reviewer 2: works?

Author: changed in commit

> In the previous version, we only had two operations that could be sent to zkSync Era from L1:

Reviewer 2: ```suggestion
In zkSync Lite, we only have two operations that could be sent to L2 from L1:
```

could --> can?

- `Deposit` to bridge funds from Ethereum to zkSync.
- `FullExit` to bridge funds back to Ethereum.

> If users wanted to deposit or withdraw funds, they had to send a transaction request to the smart contract. The request was appended to the priority queue. The queue had the following rules:

Reviewer 2: ```suggestion
If users want to deposit or withdraw funds (for withdrawals, users can request a withdrawal from L2, no using a priority transaction), they send a transaction request to the L1 smart contract. The request is appended to the priority queue. The queue had the following rules:

1. All transactions were processed sequentially, i.e. FIFO.

> 2. Each priority operation had to be processed by the operator within `X` days from submission to the contract.

Reviewer 2: ```suggestion
1. All transactions are processed sequentially, i.e. FIFO.
2. Each priority operation has to be processed by the operator within `X` days from submission to the contract.

> The first rule was strictly enforced by the smart contract. 

Reviewer 2: ```suggestion
The first rule is strictly enforced by the smart contract.

> The second rule could be violated if the operator became malicious or unavailable. If that happened, the system entered **exodus mode**, where no new blocks were processed and users could withdraw their funds without operator interference.

Reviewer 2: ```suggestion
The second rule can be violated if the operator becomes malicious or unavailable. If that happens, the system enters **exodus mode**, where no new blocks are processed (on L1) and users can withdraw their funds without operator interference.

> #### How it works in zkSync Era

Reviewer 2: I don't see it anyway, but we should state either here or on top, that this is not implemented -- it's just a design description

@kailash-manasarovar please make sure to add this before merging

Author: I removed all Priority Mode stuff.

Author: Are you saying that the whole description in the **How it works in zkSync Era** section has not been implemented?

Reviewer 2: I certainly do

Author: @barakshani I think you are confusing Priority Mode (not implemented) with Priority Queue (Yes, implemented) ... I removed all Priority Mode stuff for that reason but kept the PQ.

The previous design was sufficient for a system supporting only deposit and full exit operations. 

However, zkSync Era supports general smart contract computation which includes complex operations. For this reason, some of the principles had to evolve in order to preserve the stability of the network.

1. All transactions types are now supported by the priority queue. 

>     In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

Author: Using, instead of in?
I'm curious now about the previous statement "The first is implemented as a call to an L2 system smart contract", was this also incorrect?
@StanislavBreadless clarification please?
Yes about the branding.

>     In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

Author: ok, i removed the Lite content

>     In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

Reviewer 2: Please comment out the text, instead of deleting it, as we may use it in the future. It would be good to also have a short comment for future editors why this (commented out) bit is there.

>     In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

Author: I would rather begin again and control the process tbh. The content is still available in git.

>     In Lite, users may have had their funds locked on an L2 smart contract, and not on their own L2 account. Now, before moving funds to L1, they can execute a transaction on the zkSync Era network which releases the funds from the smart contract so they can move them.

Author: gone now

    :::info
    The withdraw functionality in zkSync Era replaces `FullExit` in Lite.
    :::

> 2. The priority queue must be fully censorship-resistant to prevent malicious activity. For example, malicious users might send multiple transactions which push up the block gas limit to unworkable levels. To mitigate against this, submitting transactions to the priority queue is no longer free and users must pay a fee to the operator. 

Author: updated in commit
